solutions/USACO feb17-circlecross.mdx

```diff
@@ -2,223 +2,91 @@
-	#endif
-}
 
-// OUTPUT
-tcT> void pr(T x) { cout << ts(x); }
-tcTUU> void pr(const T& t, const U&... u) { 
-	pr(t); pr(u...); }
-void ps() { pr("\n"); } // print w/ spaces
-tcTUU> void ps(const T& t, const U&... u) { 
-	pr(t); if (sizeof...(u)) pr(" "); ps(u...); }
-
-// DEBUG
-void DBG() { cerr << "]" << endl; }
-tcTUU> void DBG(const T& t, const U&... u) {
-	cerr << ts(t); if (sizeof...(u)) cerr << ", ";
-	DBG(u...); }
-#ifdef LOCAL // compile with -DLOCAL, chk -> fake assert
-	#define dbg(...) cerr << "Line(" << __LINE__ << ") -> [" << #__VA_ARGS__ << "]: [", DBG(__VA_ARGS__)
-	#define chk(...) if (!(__VA_ARGS__)) cerr << "Line(" << __LINE__ << ") -> function(" \
-		 << __FUNCTION__  << ") -> CHK FAILED: (" << #__VA_ARGS__ << ")" << "\n", exit(0);
-#else
-	#define dbg(...) 0
-	#define chk(...) 0
-#endif
-
-void setPrec() { cout << fixed << setprecision(15); }
-void unsyncIO() { cin.tie(0)->sync_with_stdio(0); }
-// FILE I/O
-void setIn(str s) { freopen(s.c_str(),"r",stdin); }
-void setOut(str s) { freopen(s.c_str(),"w",stdout); }
-void setIO(str s = "") {
-	unsyncIO(); setPrec();
-	// cin.exceptions(cin.failbit); 
-	// throws exception when do smth illegal
-	// ex. try to read letter into int
-	if (sz(s)) setIn(s+".in"), setOut(s+".out"); // for USACO
+int main() {
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	freopen("circlecross.in","r",stdin);
+	freopen("circlecross.out","w",stdout);
+	string s; cin >> s;
+	int ans = 0;
+	for (int a = 0; a < s.size(); ++a) 
+		for (int b = a+1; b < s.size(); ++b) 
+			for (int c = b+1; c < s.size(); ++c) 
+				for (int d = c+1; d < s.size(); ++d) 
+					ans += s[a] == s[c] && s[b] == s[d];
+	cout << ans;
+	return 0;
 }
+```
+
+Optimized $\mathcal{O}(N^2)$ solution
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
 
 int main() {
-	setIO("circlecross");
-	str s; re(s);
+	ios_base::sync_with_stdio(0); cin.tie(0);
+	freopen("circlecross.in","r",stdin);
+	freopen("circlecross.out","w",stdout);
+	string s; cin >> s;
+	int Start[26], Exit[26];
+	//Set all elements of Start and End arrays to -1.
+	memset(Start,-1,sizeof(Start));
+	memset(End,-1,sizeof(End));
+	for (int i = 0; i < s.size(); ++i) {
+		if (Start[s[i]-'A'] == -1) Start[s[i]-'A'] = i;
+		else End[s[i]-'A'] = i;
+	}
 	int ans = 0;
-	F0R(a,sz(s)) FOR(b,a+1,sz(s)) FOR(c,b+1,sz(s)) FOR(d,c+1,sz(s))
-		ans += s[a] == s[c] && s[b] == s[d];
-	ps(ans);
+	for (int i = 0; i < 26; ++i)
+		for (int j = 0; j < 26; ++j)
+			ans +=  Start[i] < Start[j] && Start[j] < End[i] && End[i] < End[j];
+	cout << ans;
+	return 0;
 }
-```
\ No newline at end of file
+```
+
+<Info title="Useful functions for initializing array values">
+
+In this problem, you can use `std::fill` instead of `memset` with no significant difference. However, with large arrays (i.e. arrays with $10^5$ or $10^6$ elements) it seems that `memset` is [more efficient](https://lemire.me/blog/2020/01/20/filling-large-arrays-with-zeroes-quickly-in-c/) than `std::fill`.
```

Comment:
Did you see what I wrote? Speed is *not* a reason to use `memset` over `fill`.

https://github.com/cpinitiative/usaco-guide/pull/658#discussion_r568318950

Plus, I don't think an array of size 10^6 doesn't qualify as large.

Reply:
So I should change `memset` into `std::fill`?

Reply:
sure